method/stLearn/method_stLearn.py

```diff
@@ -0,0 +1,235 @@
+                else:
+                    scale = 1.0
+                    spot_diameter_fullres = 50.0
+
+                adata.uns[spatial_key][library_id]["scalefactors"]["tissue_hires_scalef"] = scale
+                adata.uns[spatial_key][library_id]["scalefactors"]["spot_diameter_fullres"] = spot_dimeter_fullres
+                adata.obsm[spatial_key] = adata.obs[["imagecol", "imagerow"]].values
+                adata.obs[["imagecol", "imagerow"]] = adata.obsm["spatial"] * scale
+
+            print("Added tissue image to the object.")
+        except:
+            raise ValueError(
+                f"""\
+                {args.image!r} does not end on a valid extension.
+                """
+                )
+        # Pre-processe image
+        tile_dir = Path(out_dir / "tiles")
+        tile_dir.mkdir(parents=True, exist_ok=True)
+        # Pre-processing for spot image
+        st.pp.tiling(adata, tile_dir)
+        # This step uses deep learning model to extract high-level features from tile images
+        # May need few minutes to be completed
+        st.pp.extract_feature(adata)
+        # Convert the NumPy array to Pandas DataFrame with row names
+        embedding_df = pd.DataFrame(adata.obsm['X_morphology'], index=adata.obs_names)
+    else:
+        adata.uns["spatial"] = None
+
+    return adata
+
+
+adata = get_anndata(args)
+
+
+# Set seed
+random.seed(seed)
+np.random.seed(seed)
+
+if args.dim_red is not None:
+    # Use the provided reduced dimensionality representation.
+    adata.obsm["X_pca"] = pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+else:
+    # Calculate and use PCA
+    st.pp.scale(adata)
+    st.em.run_pca(adata, n_comps=config["n_comps"], random_state=seed)
+
+use_data = "X_pca"
+if config["image"] is True and adata.uns["spatial"] is not None:
+    if config["normalize"]:
+        # Apply stSME to normalise log transformed data
+        # With weights from morphological Similarly and physcial distance
+        st.spatial.SME.SME_normalize(adata, use_data="raw",
+                                     weights=config["weights"],
+                                     platform = technology
+                                    )
+        adata.X = adata.obsm['raw_SME_normalized']
+        st.em.run_pca(adata, n_comps=config["n_comps"])
+    else:
+        # Adjust the PCA/UMAP expression space based on identified neighbors and morphological distance by using the disk smoothing approach 
+        st.spatial.smooth.disk(adata, use_data=use_data)
+        use_data = use_data + "_disk"
+
+# Run
+if config["method"] == "kmeans":
+    st.tl.clustering.kmeans(adata,n_clusters=n_clusters, use_data=use_data, key_added="cluster", random_state=seed)
+elif config["method"] == "louvain":
+    st.pp.neighbors(adata,n_neighbors=config["n_neighbors"],use_rep=use_data, random_state=seed)
+    st.tl.clustering.louvain(adata, resolution=config["resolution"], key_added="cluster", random_state=seed)
```

Comment:
I think for other methods, we handled louvain (/leiden) by scanning the resolution to find n_clusters rather than using a config provided resolution

Reply:
> I think for other methods, we handled louvain (/leiden) by scanning the resolution to find n_clusters rather than using a config provided resolution

Was also curious about this. Found the implementation [here](https://github.com/SpatialHackathon/SpaceHack2023/blob/ac2784abacf5342b13317be7e1b7cafdb716a4cb/method/SEDR/SEDR_method.py#L123)

For Omnibenchmark, I implemented this similar R version:
`utils_tunek <- function(snn, k){
  
  for(r in seq(0.1, 5, by = 0.05)) {
    res <- igraph::cluster_leiden(graph = snn, resolution_parameter = r)
    if(res$nb_clusters >= k) break
  }
  return(res)
}
`

Reply:
I was a bit suspicious about this implementation using grid search, what if the number of clusters is outside the search range? --maybe we can keep in mind this could be the source of future errors.